File: framework-cs/apps/UnityExample/Assets/Samples/Devian Foundation/0.1.0/GamePackage/Runtime/Ability/AbilityEquipSlotPolicy.cs
using System.Collections.Generic;
using Devian.Domain.Game;

namespace Devian
{
    public enum AbilityEquipPlacementFailure
    {
        None = 0,
        SlotNotAllowed = 1,
        HandSubBlockedByTwoHandedMain = 2,
    }

    public static class AbilityEquipSlotPolicy
    {
        public static EQUIP_SLOT GetRule(EQUIP_TYPE equipType)
        {
            return TB_EQUIP_SLOT.Get(equipType);
        }

        public static bool IsAllowed(AbilityItemEquip equip, SLOT_TYPE slotType)
        {
            if (equip == null || slotType == SLOT_TYPE.NONE)
                return false;

            return IsAllowed(GetRule(equip.EquipType), slotType);
        }

        public static bool IsAllowed(EQUIP_SLOT rule, SLOT_TYPE slotType)
        {
            return rule != null
                && slotType != SLOT_TYPE.NONE
                && rule.allowed_slots != null
                && rule.allowed_slots.Contains(slotType);
        }

        public static bool IsTwoHanded(AbilityItemEquip equip)
        {
            return equip != null && IsTwoHanded(GetRule(equip.EquipType));
        }

        public static bool IsTwoHanded(EQUIP_SLOT rule)
        {
            return rule != null && rule.two_handed;
        }

        public static bool HasBlockingTwoHandedMain(IReadOnlyDictionary<SLOT_TYPE, AbilityItemEquip> equips)
        {
            if (equips == null)
                return false;

            return equips.TryGetValue(SLOT_TYPE.HAND_MAIN, out var mainHand)
                && mainHand != null
                && IsTwoHanded(mainHand);
        }

        public static AbilityEquipPlacementFailure GetPlacementFailure(
            AbilityItemEquip equip,
            SLOT_TYPE slotType,
            IReadOnlyDictionary<SLOT_TYPE, AbilityItemEquip> equips)
        {
            if (equip == null)
                return AbilityEquipPlacementFailure.SlotNotAllowed;

            return GetPlacementFailure(GetRule(equip.EquipType), slotType, equips);
        }

        public static AbilityEquipPlacementFailure GetPlacementFailure(
            EQUIP_SLOT rule,
            SLOT_TYPE slotType,
            IReadOnlyDictionary<SLOT_TYPE, AbilityItemEquip> equips)
        {
            if (!IsAllowed(rule, slotType))
                return AbilityEquipPlacementFailure.SlotNotAllowed;

            if (slotType == SLOT_TYPE.HAND_SUB && HasBlockingTwoHandedMain(equips))
                return AbilityEquipPlacementFailure.HandSubBlockedByTwoHandedMain;

            return AbilityEquipPlacementFailure.None;
        }
    }
}

File: framework-ts/module/devian-domain-game/features/ability/AbilityEquipSlotPolicy.ts
import { EQUIP_SLOT, EQUIP_TYPE, EQUIP_SLOT_TYPE, TB_EQUIP_SLOT } from '../../Generated/Game.g';
import { AbilityItemEquip } from './AbilityItemEquip';

export enum AbilityEquipPlacementFailure {
    None = 'none',
    SlotNotAllowed = 'slot_not_allowed',
    HandSubBlockedByTwoHandedMain = 'hand_sub_blocked_by_two_handed_main',
}

export class AbilityEquipSlotPolicy {
    static getRule(equipType: EQUIP_TYPE): EQUIP_SLOT | undefined {
        return TB_EQUIP_SLOT.get(equipType);
    }

    static isAllowed(equip: AbilityItemEquip | null | undefined, slotType: EQUIP_SLOT_TYPE): boolean {
        if (!equip || slotType === EQUIP_SLOT_TYPE.NONE)
            return false;

        return this.isAllowedByRule(this.getRule(equip.equipType), slotType);
    }

    static isAllowedByRule(rule: EQUIP_SLOT | null | undefined, slotType: EQUIP_SLOT_TYPE): boolean {
        return !!rule && slotType !== EQUIP_SLOT_TYPE.NONE && rule.allowed_slots.includes(slotType);
    }

    static isTwoHanded(equip: AbilityItemEquip | null | undefined): boolean {
        return !!equip && this.isTwoHandedByRule(this.getRule(equip.equipType));
    }

    static isTwoHandedByRule(rule: EQUIP_SLOT | null | undefined): boolean {
        return !!rule && rule.two_handed;
    }

    static hasBlockingTwoHandedMain(equips: ReadonlyMap<EQUIP_SLOT_TYPE, AbilityItemEquip>): boolean {
        const mainHand = equips.get(EQUIP_SLOT_TYPE.HAND_MAIN);
        return !!mainHand && this.isTwoHanded(mainHand);
    }

    static getPlacementFailure(
        equip: AbilityItemEquip | null | undefined,
        slotType: EQUIP_SLOT_TYPE,
        equips: ReadonlyMap<EQUIP_SLOT_TYPE, AbilityItemEquip>,
    ): AbilityEquipPlacementFailure {
        if (!equip)
            return AbilityEquipPlacementFailure.SlotNotAllowed;

        return this.getPlacementFailureByRule(this.getRule(equip.equipType), slotType, equips);
    }

    static getPlacementFailureByRule(
        rule: EQUIP_SLOT | null | undefined,
        slotType: EQUIP_SLOT_TYPE,
        equips: ReadonlyMap<EQUIP_SLOT_TYPE, AbilityItemEquip>,
    ): AbilityEquipPlacementFailure {
        if (!this.isAllowedByRule(rule, slotType))
            return AbilityEquipPlacementFailure.SlotNotAllowed;

        if (slotType === EQUIP_SLOT_TYPE.HAND_SUB && this.hasBlockingTwoHandedMain(equips))
            return AbilityEquipPlacementFailure.HandSubBlockedByTwoHandedMain;

        return AbilityEquipPlacementFailure.None;
    }
}
